### PR TITLE
[clang][ASTImporter] Allow import of similar friend template with different depth

### PR DESCRIPTION
**Pull request by balazske**
The fix applies to a case that occurs when the AST contains a friend template that is contained within another template and this (outer) template has specialization. (See the added test code in the commit.)
The "friend" class can have a declaration before in the AST. The friend template has the same parent as this first declaration (and is otherwise similar), but has different template depth. At AST import these declarations can be compared and import should not fail with name conflict in this case.

### CODE REVIEW

**Comment by balazske** [comment]
Currently one of the tests fail, the problem is fixed if #115518 is applied.

File: clang/lib/AST/ASTImporter.cpp [hunk 1]
```diff
@@ -6120,6 +6119,19 @@ ExpectedDecl ASTNodeImporter::VisitClassTemplateDecl(ClassTemplateDecl *D) {
...
+        // a friend template and has a specialization). For this case the
+        // declarations should match, except that the "template depth" is
+        // different. No linking of previous declaration is needed in this case.
+        // FIXME: This condition may need refinement.
+        if (D->getFriendObjectKind() != Decl::FOK_None &&
+            FoundTemplate->getFriendObjectKind() != Decl::FOK_None &&
+            D->getFriendObjectKind() != FoundTemplate->getFriendObjectKind() &&
+            IsStructuralMatch(D, FoundTemplate, /*Complain=*/false,
+                              /*IgnoreTemplateParmDepth=*/true))
+          continue;
```
Comment (Michael137) [inline clang/lib/AST/ASTImporter.cpp:6133]: It feels like the logic on 6103 and this here could be merged together? Should `IgnoreTemplateParmDepth` on 6104 just be changed to also be set if the `FriendObjectKind` matches? Then I *think* everything should work out as expected

Comment (balazske) [reply clang/lib/AST/ASTImporter.cpp:6133]: I did not want to modify an already working code part when it is not obvious that the change has no effect on the old behavior. In the new case we already know that there is no structural match and depth was not ignored, then we can check if these match with ignored depth (and this second structural equivalence check looks required to not import non-matching friends). We do not need any of the actions in the `if` branch at line 6108, only a `continue` (and it is not "dependent friend").

**Review by jcsxky — APPROVED** [review]
LGTM!

**Pull request merged**